### ISSUE TITLE
DAPO actor/loss is double-normalized: reported value depends on DP width and microbatch budget, not comparable across backends

### ISSUE THREAD

**Issue by wenchenvincent** [ZhangDanyang-AMD/Lumen-RL#26]
## Summary

`actor/loss` for DAPO is divided twice, and the second divisor is an artifact of
parallelism/packing config rather than anything about the model. The reported
number therefore changes when you change DP width or the microbatch token
budget, and it is not comparable between the FSDP2 and Megatron backends.

Training is unaffected — gradients are correct in both paths. This is a metrics
bug only. GRPO already reports the right thing; only the DAPO branch is wrong.

## Where

The inner normalization is correct and identical in both paths. `agg_loss` with
`loss_agg_mode: token-mean` returns

```python
masked_sum(loss_mat, loss_mask) / max(batch_num_tokens, 1) * dp_size
```

so each accumulation unit's loss is *already* globally normalized, and the sum
over units is the globally-normalized objective. Both paths then divide that sum
again, by different counts:

`lumenrl/engine/training/megatron_base_engine.py:512` — divides by **rows**:

```python
"loss": (
    loss_accum
    if algo_name == AlgorithmName.GRPO.value
    else loss_accum / max(1, n_rows)
),
```

`n_rows` is the number of sequences on this rank, i.e. `train_global_batch_size / DP`.

`lumenrl/workers/actor_worker.py:738` — divides by **microbatches**:

```python
if str(batch.meta.get("algorithm", "")).lower() == AlgorithmName.GRPO.value:
    metrics["loss"] = float(sum(lv))
else:
    metrics["loss"] = sum(lv) / len(lv)
```

`len(lv)` is the number of packed microbatches, i.e. roughly
`tokens_per_rank / policy.max_token_len_per_gpu`.

Both intend "mean", but over different units, and neither unit count is a
property of the optimization.

## Measurement

Two 30-step DAPO runs, Qwen3-8B-Base, vLLM rollout, **identical config and seed
10086** — `dapo_qwen3_8b_ray_megatron_smoke.yaml` (megatron_native, DP=2) versus
`dapo_qwen3_8b_ray_vllm_smoke.yaml` (fsdp2, DP=8). Every hyperparameter that
affects training matches: lr 1e-6, warmup 10, global batch 128, gen batch 24,
16 generations, clip 0.2/0.28/10, max response 512.

| metric | megatron_native | fsdp2 | ratio |
|---|---|---|---|
| `rollout_corr/kl` | 0.00123 | 0.00123 | 1.00x |
| `rollout_corr/kl` drift (last10-first10) | +0.00014 | +0.00014 | — |
| `ppl_ratio` | 1.00185 | 1.00161 | 1.00x |
| `IS_ESS` | 0.99757 | 0.99764 | 1.00x |
| `actor/ppo_kl` | +1e-5 | -9e-5 | both ~0 |
| `actor/entropy` | 0.6556 | 0.6672 | 0.98x |
| `actor/grad_norm` | 1.029 | 1.135 | 0.91x |
| `reward/accuracy` | 0.1401 | 0.1297 | 1.08x |
| tokens/step | 61915 | 59248 | 1.05x |
| **`actor/loss`** | **0.00359** | **0.06256** | **0.06x** |

Everything agrees within 9% except the loss, which differs 17x.

The divisors account for it:

```
megatron divisor  n_rows          = 128/2                   = 64
fsdp     divisor  n_microbatches  = (59248/8)/2048          ~ 3.62

predicted ratio   (1/64)/(1/3.62) = 0.0565
observed  ratio   0.00359/0.06256 = 0.0575      <- 2% off
```

Undoing each divisor recovers the same objective:

```
megatron  0.00359 x 64    = 0.2301
fsdp      0.06256 x 3.62  = 0.2262     <- agree within 2%
```

2% is tighter than any other metric in the table, which is the point: the two
backends optimize the same objective, and only the reported scalar differs.
Consistent with `grad_norm` matching within 9% — each `backward()` accumulates
the un-divided per-unit loss, so gradients were never affected.

## Impact

* `actor/loss` is not comparable between fsdp2 and megatron_native.
* Within one backend it is not comparable across DP widths (megatron) or across
  `max_token_len_per_gpu` values (fsdp).
* A loss curve that looks 17x different between two backends invites exactly the
  wrong conclusion during a backend migration. It cost us a round of
  investigation before the divisors turned up.

## Suggested fix

Make DAPO match GRPO, which is already correct — report the sum, dropping the
extra mean:

```python
# megatron_base_engine.py
"loss": loss_accum,

# actor_worker.py
metrics["loss"] = float(sum(lv))
```

Worth a regression test asserting the reported loss is invariant to DP width and
to `max_token_len_per_gpu` on a fixed batch — that is the property being violated
and it is cheap to pin.

One caveat for whoever picks this up: the fix changes the scale of every
historical DAPO loss curve (by `n_rows` or `n_microbatches`), so dashboards and
any saved baselines shift. That is a call for the team, not obviously a
do-it-quietly change. Everything else — `ppo_kl`, `grad_norm`, `entropy`,
`rollout_corr/*` — is unaffected.

## Environment

Reproduced on `main` @ 7e9c16b; both call sites are present there unchanged.
8x MI355X (gfx950), ROCm, Megatron-Core 0.18.2, TE 2.15.0.dev0+6e541a1, vLLM 0.23.0.

Found while collecting cross-backend acceptance evidence for #25; unrelated to
that PR's changes.

**Comment by wenchenvincent** [comment]
Fix proposed in #29.

Both call sites now route through a single `reduce_reported_loss` in `lumenrl/algorithms/loss_functions.py` — sharing the reduction is part of the fix, since the two paths had separate inline reductions and that is how they came to disagree.

One correction to what I suggested above: I first wrote the regression test against `agg_loss`, and it **passed on the unfixed tree**. `agg_loss` was never broken — the reporting was — so an `agg_loss`-only test cannot catch this. #29's tests assert the reduction itself and that neither call site re-implements it, and are falsified by reintroducing either divisor.

The dashboard-scale caveat in the last section still applies and is called out for reviewers; if the team prefers preserving the current scale and only fixing the cross-backend inconsistency, that is a different patch and I can do it instead.
